### PR TITLE
[kotlin] Prohibit overriding equals/hashCode methods for value classes

### PR DESCRIPTION
**Pull request by t-kameyama**
https://github.com/JetBrains/intellij-community/pull/2126#issuecomment-1229211152

### CODE REVIEW

File: plugins/kotlin/fir/test/org/jetbrains/kotlin/idea/fir/codeInsight/FirOverrideImplementTest.kt [hunk 1]
```diff
@@ -198,6 +198,10 @@ internal class FirOverrideImplementTest : OverrideImplementTest<KtClassMember>()
         super.testNoAnyMembersInInterface()
     }
 
+    override fun testNoAnyMembersInValueClass() {
+        super.testNoAnyMembersInValueClass()
+    }
```
Comment (t-kameyama) [inline plugins/kotlin/fir/test/org/jetbrains/kotlin/idea/fir/codeInsight/FirOverrideImplementTest.kt:203]: This test fails.
```
org.jetbrains.kotlin.utils.KotlinExceptionWithAttachments: Error while resolving org.jetbrains.kotlin.fir.declarations.impl.FirSimpleFunctionImpl 
from RAW_FIR to BODY_RESOLVE
current declaration phase TYPES
declaration origin: Source
declaration session: class org.jetbrains.kotlin.analysis.low.level.api.fir.sessions.LLFirSourcesSession
declaration module data: class org.jetbrains.kotlin.analysis.low.level.api.fir.project.structure.LLFirModuleData
declaration KtModule: class org.jetbrains.kotlin.idea.base.projectStructure.KtSourceModuleByModuleInfo
declaration platform: JVM (1.8)


	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.LLFirModuleLazyDeclarationResolverKt.rethrowWithDetails(LLFirModuleLazyDeclarationResolver.kt:464)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.LLFirModuleLazyDeclarationResolverKt.access$rethrowWithDetails(LLFirModuleLazyDeclarationResolver.kt:1)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.LLFirModuleLazyDeclarationResolver.lazyResolveDeclaration(LLFirModuleLazyDeclarationResolver.kt:245)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.file.structure.FileStructure.createDeclarationStructure(FileStructure.kt:105)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.file.structure.FileStructure.createStructureElement(FileStructure.kt:130)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.file.structure.FileStructure.access$createStructureElement(FileStructure.kt:21)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.file.structure.FileStructure$getStructureElementForDeclaration$structureElement$1.invoke(FileStructure.kt:49)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.file.structure.FileStructure$getStructureElementForDeclaration$structureElement$1.invoke(FileStructure.kt:47)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.file.structure.FileStructure.getStructureElementForDeclaration$lambda$0(FileStructure.kt:47)
	at java.base/java.util.concurrent.ConcurrentHashMap.compute(ConcurrentHashMap.java:1916)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.file.structure.FileStructure.getStructureElementForDeclaration(FileStructure.kt:47)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.file.structure.FileStructure.getStructureElementFor(FileStructure.kt:42)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.element.builder.FirElementBuilder.getOrBuildFirForNonKtFileElement(FirElementBuilder.kt:93)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.element.builder.FirElementBuilder.getOrBuildFirFor(FirElementBuilder.kt:66)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.state.LLFirResolvableResolveSession.getOrBuildFirFor$low_level_api_fir(LLFirResolvableResolveSession.kt:57)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.api.LowLevelFirApiFacadeKt.getOrBuildFir(LowLevelFirApiFacade.kt:108)
	at org.jetbrains.kotlin.analysis.api.fir.components.KtFirReferenceShortener.collectShortenings(KtFirReferenceShortener.kt:666)
	at org.jetbrains.kotlin.analysis.api.components.KtReferenceShortenerMixIn$DefaultImpls.collectPossibleReferenceShortenings(KtReferenceShortener.kt:69)
	at org.jetbrains.kotlin.analysis.api.KtAnalysisSession.collectPossibleReferenceShortenings(KtAnalysisSession.kt:29)
	at org.jetbrains.kotlin.analysis.api.components.KtReferenceShortenerMixIn$DefaultImpls.collectPossibleReferenceShortenings$default(KtReferenceShortener.kt:62)
	at org.jetbrains.kotlin.idea.core.overrideImplement.KtGenerateMembersHandler.generateMembers(KtGenerateMembersHandler.kt:56)
	at org.jetbrains.kotlin.idea.codeInsight.AbstractOverrideImplementTest$performGenerateCommand$1.invoke(AbstractOverrideImplementTest.kt:163)
	at org.jetbrains.kotlin.idea.codeInsight.AbstractOverrideImplementTest$performGenerateCommand$1.invoke(AbstractOverrideImplementTest.kt:162)
	at org.jetbrains.kotlin.idea.util.application.ApplicationUtilsKt.runWriteAction$lambda-1(ApplicationUtils.kt:24)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$runWriteAction$12(ApplicationImpl.java:983)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteActionWithClass(ApplicationImpl.java:962)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:983)
	at org.jetbrains.kotlin.idea.util.application.ApplicationUtilsKt.runWriteAction(ApplicationUtils.kt:24)
	at org.jetbrains.kotlin.idea.util.application.ApplicationUtilsKt.executeWriteCommand$lambda-3(ApplicationUtils.kt:42)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:219)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:174)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:164)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:150)
	at org.jetbrains.kotlin.idea.util.application.ApplicationUtilsKt.executeWriteCommand(ApplicationUtils.kt:42)
	at org.jetbrains.kotlin.idea.codeInsight.AbstractOverrideImplementTest.performGenerateCommand(AbstractOverrideImplementTest.kt:162)
	at org.jetbrains.kotlin.idea.codeInsight.AbstractOverrideImplementTest.doMultiOverrideImplement(AbstractOverrideImplementTest.kt:140)
	at org.jetbrains.kotlin.idea.codeInsight.AbstractOverrideImplementTest.doMultiFileTest(AbstractOverrideImplementTest.kt:95)
	at org.jetbrains.kotlin.idea.codeInsight.AbstractOverrideImplementTest.doMultiOverrideFileTest(AbstractOverrideImplementTest.kt:51)
	at org.jetbrains.kotlin.idea.codeInsight.OverrideImplementTest.testNoAnyMembersInValueClass(OverrideImplementTest.kt:199)
	at org.jetbrains.kotlin.idea.fir.codeInsight.FirOverrideImplementTest.testNoAnyMembersInValueClass(FirOverrideImplementTest.kt:202)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at com.intellij.testFramework.UsefulTestCase.lambda$runBare$10(UsefulTestCase.java:475)
	at com.intellij.testFramework.UsefulTestCase.lambda$wrapTestRunnable$12(UsefulTestCase.java:496)
	at com.intellij.testFramework.UsefulTestCase.runTestRunnable(UsefulTestCase.java:400)
	at com.intellij.testFramework.UsefulTestCase.defaultRunBare(UsefulTestCase.java:416)
	at com.intellij.testFramework.UsefulTestCase.lambda$runBare$11(UsefulTestCase.java:483)
	at com.intellij.testFramework.EdtTestUtil.lambda$runInEdtAndWait$1(EdtTestUtil.java:40)
	at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:209)
	at com.intellij.openapi.application.TransactionGuardImpl.access$100(TransactionGuardImpl.java:21)
	at com.intellij.openapi.application.TransactionGuardImpl$1.run(TransactionGuardImpl.java:191)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:832)
	at com.intellij.openapi.application.impl.ApplicationImpl$3.run(ApplicationImpl.java:448)
	at com.intellij.openapi.application.impl.LaterInvocator$1.run(LaterInvocator.java:97)
	at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:73)
	at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:116)
	at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:36)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:779)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:730)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:724)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:749)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:916)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:764)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:447)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:773)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:446)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:113)
	at com.intellij.ide.IdeEventQueue.performActivity(IdeEventQueue.java:622)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:444)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:832)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:490)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
Caused by: java.lang.IllegalStateException: Expected FirResolvedTypeRef with ConeKotlinType but was FirUserTypeRefImpl Int
	at org.jetbrains.kotlin.fir.types.FirTypeUtilsKt.getConeType(FirTypeUtils.kt:33)
	at org.jetbrains.kotlin.fir.declarations.ValueClassesUtilsKt.computeValueClassRepresentation(ValueClassesUtils.kt:41)
	at org.jetbrains.kotlin.fir.resolve.transformers.FirStatusResolveTransformer.transformRegularClass(FirStatusResolveTransformer.kt:115)
	at org.jetbrains.kotlin.fir.resolve.transformers.FirStatusResolveTransformer.transformRegularClass(FirStatusResolveTransformer.kt:79)
	at org.jetbrains.kotlin.fir.declarations.FirRegularClass.transform(FirRegularClass.kt:50)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.transformers.LLFirDesignatedStatusResolveTransformer$FirDesignatedStatusResolveTransformerForIDE.transformDeclarationContent(LLFirDesignatedStatusResolveTransformer.kt:79)
	at org.jetbrains.kotlin.fir.resolve.transformers.AbstractFirStatusResolveTransformer.transformFile(FirStatusResolveTransformer.kt:268)
	at org.jetbrains.kotlin.fir.resolve.transformers.AbstractFirStatusResolveTransformer.transformFile(FirStatusResolveTransformer.kt:252)
	at org.jetbrains.kotlin.fir.declarations.FirFile.transform(FirFile.kt:42)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.transformers.LLFirDesignatedStatusResolveTransformer$transformDeclaration$1$invoke$$inlined$runPhaseWithCustomResolve$1.run(utils.kt:78)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeNonCancelableSection$3(CoreProgressManager.java:224)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:665)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:621)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$computeInNonCancelableSection$4(CoreProgressManager.java:232)
	at com.intellij.openapi.progress.Cancellation.computeInNonCancelableSection(Cancellation.java:99)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeInNonCancelableSection(CoreProgressManager.java:232)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeNonCancelableSection(CoreProgressManager.java:223)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.transformers.LLFirDesignatedStatusResolveTransformer$transformDeclaration$1.invoke(LLFirDesignatedStatusResolveTransformer.kt:72)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.transformers.LLFirDesignatedStatusResolveTransformer$transformDeclaration$1.invoke(LLFirDesignatedStatusResolveTransformer.kt:45)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.ResolveTreeBuilder$resolvePhase$1$1.invoke(ResolveTreeBuilder.kt:47)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.ResolveTreeBuilder$resolvePhase$1$1.invoke(ResolveTreeBuilder.kt:46)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.ResolveTreeBuilder.resolveToDeclaration(ResolveTreeBuilder.kt:117)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.ResolveTreeBuilder.resolvePhase(ResolveTreeBuilder.kt:46)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.transformers.LLFirDesignatedStatusResolveTransformer.transformDeclaration(LLFirDesignatedStatusResolveTransformer.kt:45)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.LLFirModuleLazyDeclarationResolver.runLazyDesignatedResolveWithoutLock(LLFirModuleLazyDeclarationResolver.kt:384)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.LLFirModuleLazyDeclarationResolver.access$runLazyDesignatedResolveWithoutLock(LLFirModuleLazyDeclarationResolver.kt:37)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.LLFirModuleLazyDeclarationResolver$doLazyResolveDeclaration$1$1.invoke(LLFirModuleLazyDeclarationResolver.kt:345)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.LLFirModuleLazyDeclarationResolver$doLazyResolveDeclaration$1$1.invoke(LLFirModuleLazyDeclarationResolver.kt:344)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.ResolveTreeBuilder$resolveEnsure$1$1.invoke(ResolveTreeBuilder.kt:60)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.ResolveTreeBuilder$resolveEnsure$1$1.invoke(ResolveTreeBuilder.kt:59)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.ResolveTreeBuilder.resolveToDeclaration(ResolveTreeBuilder.kt:117)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.ResolveTreeBuilder.resolveEnsure(ResolveTreeBuilder.kt:59)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.LLFirModuleLazyDeclarationResolver.doLazyResolveDeclaration(LLFirModuleLazyDeclarationResolver.kt:344)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.lazy.resolve.LLFirModuleLazyDeclarationResolver.lazyResolveDeclaration(LLFirModuleLazyDeclarationResolver.kt:243)
	... 80 more
```

Comment (vladimirdolzhenko) [reply plugins/kotlin/fir/test/org/jetbrains/kotlin/idea/fir/codeInsight/FirOverrideImplementTest.kt:203]: I'd consider to add `IGNORE_FIR` to test data

File: plugins/kotlin/code-insight/override-implement-k2/src/org/jetbrains/kotlin/idea/core/overrideImplement/KtOverrideMembersHandler.kt [hunk 2]
```diff
@@ -82,6 +80,14 @@ open class KtOverrideMembersHandler : KtGenerateMembersHandler(false) {
                                 BodyType.NoBody
                             }
                         }
+                        (classOrObjectSymbol as? KtNamedClassOrObjectSymbol)?.isInline == true &&
+                                containingSymbol?.classIdIfNonLocal == StandardClassIds.Any -> {
+                            if ((symbolToProcess as? KtFunctionSymbol)?.name?.asString() in listOf("equals", "hashCode")) {
```
Comment (vladimirdolzhenko) [inline plugins/kotlin/code-insight/override-implement-k2/src/org/jetbrains/kotlin/idea/core/overrideImplement/KtOverrideMembersHandler.kt:85]: minor: `symbolToProcess.safeAs<KtFunctionSymbol>?.name`

File: plugins/kotlin/fir/test/org/jetbrains/kotlin/idea/fir/codeInsight/FirOverrideImplementTest.kt (hunk 1 above)
Comment (vladimirdolzhenko) [reply plugins/kotlin/fir/test/org/jetbrains/kotlin/idea/fir/codeInsight/FirOverrideImplementTest.kt:203]: @fedochet confirmed - the issue is on our side, 

so, I would not block this PR until the issue will be fixed and prefer to use `IGNORE_FIR` and merge MR rather wait the fix

Comment (JAbduvohidov) [reply plugins/kotlin/fir/test/org/jetbrains/kotlin/idea/fir/codeInsight/FirOverrideImplementTest.kt:203]: ҷураҷон кодт маъқул нашид

Comment (fedochet) [reply plugins/kotlin/fir/test/org/jetbrains/kotlin/idea/fir/codeInsight/FirOverrideImplementTest.kt:203]: I've filed the issue: https://youtrack.jetbrains.com/issue/KTIJ-22630

The fix is on it's way 🚂

Comment (fedochet) [reply plugins/kotlin/fir/test/org/jetbrains/kotlin/idea/fir/codeInsight/FirOverrideImplementTest.kt:203]: The issue is fixed 🎉 

https://youtrack.jetbrains.com/issue/KTIJ-22630

**Comment by vladimirdolzhenko** [comment]
LGTM, merged, thanks

**Pull request closed without merging**